### PR TITLE
Requirements schema

### PR DESCRIPTION
**Pull request by rafaelfolco**
Validate the requirements JSON against a schema file.

### CODE REVIEW

**Comment by rafaelfolco** [comment]
unit tests: in progress (to be added in separate commits, same PR)

File: JSON/req-schema.json [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                        "$ref": "#/definitions/mv_param"
+                    }
+                }
+            },
+            "additionalProperties": false
+        },
+        "validations": {
+            "type": "object",
+            "patternProperties": {
+                "^[A-Za-z]+[A-Za-z0-9_-]*$": {
```
Comment (k-rister) [inline JSON/req-schema.json:24]: Just curious...are you saying that for the validations the property name has to start with a letter?  If so, why?

File: JSON/req-schema.json [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                    "$ref": "#/definitions/validation_obj"
+                }
+            },
+            "additionalProperties": false
+        },
+        "units": {
+            "type": "object",
+            "patternProperties": {
+                "^[A-Za-z]+[A-Za-z0-9_-]*$": {
```
Comment (k-rister) [inline JSON/req-schema.json:33]: same question

File: JSON/req-schema.json [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                    "$ref": "#/definitions/unit_obj"
+                }
+            },
+            "additionalProperties": false
+        }
+    },
+    "required": [ "validations" ],
```
Comment (k-rister) [inline JSON/req-schema.json:40]: should there be an `"additionalProperties": false` here?

File: JSON/req-schema.json [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                },
+                "description": {
+                    "type": "string",
+                    "minLength": 1
+                },
+                "convert": {
+                    "type": "string",
+                    "minLength": 1
+                },
+                "transform": {
```
Comment (k-rister) [inline JSON/req-schema.json:101]: you need an `"additionalProperties": false` here also...you need it on every object you create if you don't want arbitrary objects to be created by the user.

File: JSON/req-schema.json [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            },
+            "required": [
+                "args",
+                "vals"
+            ],
+            "additionalProperties": false
+        },
+        "unit_obj": {
+            "type": "object",
+            "minItems": 0,
```
Comment (k-rister) [inline JSON/req-schema.json:122]: does it make sense to create a `unit_obj` that is empty?  shouldn't that case be covered by making the `unit_obj` optional above?

File: JSON/req-schema.json (hunk 1 above)
Comment (rafaelfolco) [reply JSON/req-schema.json:24]: Is there any param called "99frame-size" or "1st-param-name" ? With this pattern you can have numbers, but not at the beginning. I can change if this is a requirement. Also, all the regex patterns in the json files can be easily changed if you need (even locally without a PR).

File: JSON/req-schema.json (hunk 2 above)
Comment (rafaelfolco) [reply JSON/req-schema.json:33]: again, I can change it as you wish

File: JSON/req-schema.json (hunk 3 above)
Comment (rafaelfolco) [reply JSON/req-schema.json:40]: yes, accidentally removed it

File: JSON/req-schema.json (hunk 1 above)
Comment (k-rister) [reply JSON/req-schema.json:24]: I think we can't predict what the parameter names are so we should try to be as flexible as possible.

File: JSON/req-schema.json (hunk 4 above)
Comment (rafaelfolco) [reply JSON/req-schema.json:101]: done

File: JSON/req-schema.json (hunk 5 above)
Comment (rafaelfolco) [reply JSON/req-schema.json:122]: `units` is an object, not an array, so this is wrong. Changing to:
```
"minProperties": 1
```

File: JSON/req-schema.json (hunk 1 above)
Comment (rafaelfolco) [reply JSON/req-schema.json:24]: ack

File: JSON/req-schema.json (hunk 5 above)
Comment (k-rister) [reply JSON/req-schema.json:122]: oops...missed that part myself

**Comment by rafaelfolco** [comment]
WIP: tests

**Comment by rafaelfolco** [comment]
will add tests and fix corner cases if needed in a follow-up PR

**Review by k-rister — APPROVED** [review]
